source/real_locations.rst

```diff
@@ -0,0 +1,116 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We can solve the problem by adding an attribute called ``currentLocation`` to the ``Person`` class.
+
+.. tip:: Remember to also initialize the attribute when created an instance of the class while extracting data from the CSV!
+
+After doing so, we need to add a behaviour which changes the ``currentLocation`` at every tick. We could do this by adding the following line of code in the ``Person`` class:
+
+.. code-block:: scala
+
+  private val checkCurrentLocation: Context => Unit = (context: Context) => {
+    val schedule = context.fetchScheduleFor(this).get
+    val locationNextTick: String = schedule.getForStep(context.getCurrentStep + 1)
+    if (currentLocation != locationNextTick) {
+      this.updateParam("currentLocation", locationNextTick)
+    }
+  }
+  addBehaviour(checkCurrentLocation)
+
+.. hint:: ``updateParam`` only updates the value of the attribute at the **end** of the tick. Thus, for all practical purposes, it's useful to view the function as one that changes the value of the attribute on the *subsequent tick*. As such, we store the place the person is expected to be on the next tick, and hencee use ``context.getCurrentStep+1`` as an argument to ``schedule.getForStep``.
+
+Now, we need use this attribute when we compute ``totalNeighbourCount`` and ``infectedNeighbourCount``. ``node.getConnectionCount`` has another (optional) argument besides the relation, which is ``matchPattern``. Using it, we can get counts of the people with the relationship who's attributes follow a pattern: in this case, we'll look for the people who have the ``currentLocation`` attribute equal to the ``placeType`` of the node.
+
+
+.. code-block:: scala
+
+  def computeInfectedFraction(node: Node, placeType: String, context: Context): Double = {
+    val totalNeighbourCount = node.getConnectionCount(node.getRelation[Person]().get,
+      "currentLocation" equ placeType)
+
+    if (totalNeighbourCount == 0) return 0d
+
+    val infectedNeighbourCount = node.getConnectionCount(node.getRelation[Person]().get,
+      ("infectionState" equ Infected) and ("currentLocation" equ placeType))
+
+    infectedNeighbourCount.toDouble / totalNeighbourCount.toDouble
+  }
+
+.. note:: ``equ``, ``and`` and other pattern-matching relations are defined in ``import com.bharatsim.engine.graph.patternMatcher.MatchCondition._``
+
+Checking the locations without a parameter
+------------------------------------------
+
+``updateParam`` updates a node on the graph, and is called once per person per tick. That can potentially slow the program down, and another possibility is to avoid using it entirely. We'll still do the same thing - get the schedule for the agent, check if they're actually at the place you're looking at, and then get the total and infected counts.
+
+The following function accomplishes this:
+
+.. code-block:: scala
+
+  def computeInfectedFraction(node: Node, placeType: String, context: Context): Double = {
```

Comment:
Put some spaces in this code block, keep the chunks that are "related" next to each other, and add spaces to keep them distinct for a reader.

Reply:
Actually, this code is a little harder to understand than the previous one. Could you break it up into bits and explain it as you did for the code in the "Inputs and Outputs" section? Link to the Scala documentation on iterators as well.

Reply:
I'll do that, will iterators be explained in the documentation or do you want me to link to an alvin-alexander sort of page?

Reply:
Keep it external. Doesn't seem like the sort of thing we'd be writing too much about.